### PR TITLE
Template rework and Unused Check

### PR DESCRIPTION
**Pull request by RobAndrewHurst**
## Description

I have updated the test function on the workspace mod to include a list of used templates with their frequency as well as a list of templates that are not used at all.

I also add the _core property on all default templates that get assigned to the workspace.templates object.

I also noticed that the template property on a layer when merging a template into that layer would overwrite the template on the layer. To combat this I delete the template property on the template before merging into the layer.

## GitHub Issue
#1494 
## Type of Change
Please delete options that are not relevant, and select all options that apply. 

- ✅ Bug fix (non-breaking change which fixes an issue)
- ✅ New feature (non-breaking change which adds functionality)
- ✅ Testing

### CODE REVIEW

**Comment by dbauszus-glx** [comment]
I added the logic to mark a template object as core to a helper function so that this can be described and doesn't have to be repeated multiple times in the workspace.templates assign.

The custom_templates should not be marked as core.

```js
  /**
  @function mark_core

  @description
  The method maps the Object.entries of the templates_object param and assigns the _core: true property to the object marking this as a system template.

  @param {Object} templates_object 
  @returns {Object} templates_object with _core: true property.
  */
  function mark_core(templates_object) {

    return Object.fromEntries(
      Object.entries(templates_object)
        .map(([key, template]) => [key, { ...template, _core: true }])
    )
  }

  // Assign default view and query templates to workspace.
  workspace.templates = {

    ...mark_core(view_templates),
    ...mark_core(mail_templates),
    ...mark_core(msg_templates),
    ...mark_core(query_templates),

    ...custom_templates,

    // Default templates can be overridden by assigning a template with the same key.
    ...workspace.templates
  }
```

File: mod/workspace/cache.js [hunk 1]
```diff
@@ -95,25 +95,33 @@ async function cacheWorkspace() {
...
+  /**
+  @function mark_core
+
+  @description
+  The method maps the Object.entries of the templates_object param and assigns the _core: true property to the object marking this as a system template.
+
+  @param {Object} templates_object 
+  @returns {Object} templates_object with _core: true property.
+  */
+  function mark_core(templates_object) {
```
Comment (RobAndrewHurst) [inline mod/workspace/cache.js:107]: Nice, good use of a method to mark them.

**Comment by RobAndrewHurst** [comment]
image
The templates that are core are not being marked as core now.

**Comment by RobAndrewHurst** [comment]
> image The templates that are core are not being marked as core now.

Oh wait, because they can be. Because these are coming from custom templates.

**Review by dbauszus-glx — DISMISSED** [review]
The checkTemplate methods needs to be documented as it isn't clear what this is doing.

image

The check fails on this test workspace. OSM template is used.

```json
{
  "templates": {
    "OSM": {
      "test": true
    }
  },
  "locale": {
    "layers": {
      "OSM": {
        "template": "OSM",
        "display": true,
        "format": "tiles",
        "URI": "https://{a-c}.tile.openstreetmap.org/{z}/{x}/{y}.png",
        "attribution": {
          "© OpenStreetMap": "http://www.openstreetmap.org/copyright"
        }
      }
    }
  }
}
```

image

**Comment by dbauszus-glx** [comment]
@RobAndrewHurst It is possible to overwrite _core templates like with custom templates. But the overwritten ones should not be marked as core. The _core flag is only to indicate system templates which come from the templates directory itself.

**Comment by RobAndrewHurst** [comment]
> @RobAndrewHurst It is possible to overwrite _core templates like with custom templates. But the overwritten ones should not be marked as core. The _core flag is only to indicate system templates which come from the templates directory itself.

Yeah I agree with this.

**Comment by RobAndrewHurst** [comment]
> @RobAndrewHurst It is possible to overwrite _core templates like with custom templates. But the overwritten ones should not be marked as core. The _core flag is only to indicate system templates which come from the templates directory itself.

I have added in different types on the entries added to the workspace.templates.

The core templates will have 'core'. custom_templates will be 'custom_templates' & templates that come from the workspace itself be it from a layer itself or just in the workspace.templates object itself will have 'workspace_templates'.

I then filter out the workspace_templates as those are the templates only required for the testing, but with these additional '_type' properties we can do further drilling. 

Please have a look, and let me know if you agree with this or if we can apply this can of logic in a more elegant way.

**Comment by dbauszus-glx** [comment]
> > @RobAndrewHurst It is possible to overwrite _core templates like with custom templates. But the overwritten ones should not be marked as core. The _core flag is only to indicate system templates which come from the templates directory itself.
> 
> I have added in different types on the entries added to the workspace.templates.
> 
> The core templates will have 'core'. custom_templates will be 'custom_templates' & templates that come from the workspace itself be it from a layer itself or just in the workspace.templates object itself will have 'workspace_templates'.
> 
> I then filter out the workspace_templates as those are the templates only required for the testing, but with these additional '_type' properties we can do further drilling.
> 
> Please have a look, and let me know if you agree with this or if we can apply this can of logic in a more elegant way.

I like that. Much better to use _type with a string value.

Can you add a global template typedef with all the properties eg. template, _type, src, err, etc.

**Review by simon-leech — APPROVED** [review]
I'm happy with this - easily explainable and the tests are working well to help clean up!

**Review by AlexanderGeere — APPROVED** [review]
Nice One

**Pull request merged**